### PR TITLE
Use #nonnull directive to determine NonNullTypes context

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@AlekseyTs, @dotnet/roslyn-compiler please review.

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 1]
```diff
@@ -122,12 +122,12 @@ protected static CSharpCompilationOptions WithNonNullTypesFalse(CSharpCompilatio
 
         protected static string NonNullTypesOff()
         {
-            return $"#pragma warning disable {(int)ErrorCode.WRN_PragmaNonNullTypes}";
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs]: >WRN_PragmaNonNullTypes [](start = 61, length = 22)

Can we remove this warning now?

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 2]
```diff
@@ -866,31 +866,16 @@ public virtual bool HasUnsupportedMetadata
...
-                        if (state.GetValueOrDefault() == ReportDiagnostic.Suppress)
-                        {
-                            result = false;
-                        }
-                        else
-                        {
-                            return true;
-                        }
-                    }
+                    return ((CSharpSyntaxTree)tree).GetNonNullDirectiveState(location.SourceSpan.Start);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs]: >return ((CSharpSyntaxTree)tree).GetNonNullDirectiveState(location.SourceSpan.Start); [](start = 20, length = 84)

It looks like semantics of merging across different partial type declaration is not preserved. The lack of tests that catch this is probably my fault.

File: src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs [hunk 3]
```diff
@@ -84,8 +84,11 @@ private BoundExpression BindAnonymousObjectCreation(AnonymousObjectCreationExpre
 
                 // build anonymous type field descriptor
                 fieldSyntaxNodes[i] = (nameToken.Kind() == SyntaxKind.IdentifierToken) ? (CSharpSyntaxNode)nameToken.Parent : fieldInitializer;
-                fields[i] = new AnonymousTypeField(fieldName == null ? "$" + i.ToString() : fieldName, fieldSyntaxNodes[i].Location, 
-                                                   TypeSymbolWithAnnotations.Create(fieldType, isNullableIfReferenceType: node.IsFeatureStaticNullCheckingEnabled()));
+                bool? isNullableIfReferenceType = ((CSharpParseOptions)node.SyntaxTree?.Options)?.IsFeatureEnabled(MessageID.IDS_FeatureStaticNullChecking) == true;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs]: >isNullableIfReferenceType [](start = 22, length = 25)

Not related to this PR: I don't think we should bind differently based on feature flag. We should probably always bind anonymous type members as null-oblivious and then consider revising during nullability analysis.
Could you update this code with a comment and a link to [https://github.com/dotnet/roslyn/issues/24018](https://github.com/dotnet/roslyn/issues/24018) ?

File: src/Compilers/CSharp/Portable/Syntax/NonNullDirectiveMap.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        /// <summary>
+        /// Returns true if the `#nonnull` directive preceding the position is
+        /// `restore`, false for if `disable`, and null if no preceding directive.
+        /// </summary>
+        internal bool? GetDirectiveState(int position)
+        {
+            int index = _directives.BinarySearch((position, false), PositionComparer.Instance);
+            if (index < 0)
+            {
+                index = ~index - 1;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/NonNullDirectiveMap.cs]: >index = ~index - 1; [](start = 16, length = 19)

I didn't understand the purpose of this.

Comment (cston) [reply src/Compilers/CSharp/Portable/Syntax/NonNullDirectiveMap.cs]: If there is not an exact match, `BinarySearch` returns the complement of the nearest index.

---
In reply to: [225352180](https://github.com/dotnet/roslyn/pull/30499#discussion_r225352180) [](ancestors = 225352180)

**Comment by cston** [comment]
@dotnet/bot test windows_debug_unit32_prtest

**Comment by cston** [comment]
@dotnet-bot test windows_debug_unit32_prtest

**Pull request merged**